### PR TITLE
[pcl_ros] Update pcd file to publish by service call

### PR DESCRIPTION
**Pull request by wkentaro**

### CODE REVIEW

**Comment by wkentaro** [comment]
ping

**Comment by wkentaro** [comment]
ping

File: pcl_ros/CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ find_package(PCL REQUIRED)
 find_package(catkin REQUIRED COMPONENTS
   dynamic_reconfigure
   genmsg
+  message_generation
```
Comment (paulbovbel) [inline pcl_ros/CMakeLists.txt]: Shouldn't generate messages here. There's a pcl_msgs package for this purpose

File: pcl_ros/tools/pcd_to_pointcloud.cpp [hunk 2]
```diff
@@ -89,46 +94,48 @@ class PCDGenerator
...
+        /*oneshot=*/false);
+      // service to update filename
+      service_ = private_nh_.advertiseService("update_filename", &PCDGenerator::updateFilenameCb, this);
       return (0);
     }
 
+    bool updateFilenameCb(pcl_ros::UpdateFilename::Request &req,
+                          pcl_ros::UpdateFilename::Response &res)
+    {
+      boost::mutex::scoped_lock lock(mutex_);
```
Comment (paulbovbel) [inline pcl_ros/tools/pcd_to_pointcloud.cpp]: No need to mutex - this node is single threaded

File: pcl_ros/tools/pcd_to_pointcloud.cpp [hunk 3]
```diff
@@ -89,46 +94,48 @@ class PCDGenerator
       if (file_name_ == "" || pcl::io::loadPCDFile (file_name_, cloud_) == -1)
         return (-1);
       cloud_.header.frame_id = tf_frame_;
+      // set timer to publish cloud periodically
+      timer_ = nh_.createWallTimer(
```
Comment (paulbovbel) [inline pcl_ros/tools/pcd_to_pointcloud.cpp]: preference to timer would be to convert publisher to latched, and make service call trigger publishing the new point cloud, much simpler flow

**Comment by paulbovbel** [comment]
@wkentaro, sorry for the delay, and thanks for the contribution. I've made some comments.

**Comment by wkentaro** [comment]
@paulbovbel 
Thanks for the comments.
I have added some modifying commits. 
Could you please review them? then I will squash the commits.

**Comment by paulbovbel** [comment]
Thanks for addressing the message issue.

With regards to the republish timer, this would probably be more straightforward with a latched publisher, which would update the latched message every time the service is called. Would that work for your application?

**Comment by wkentaro** [comment]
> With regards to the republish timer, this would probably be more straightforward with a latched publisher, which would update the latched message every time the service is called. Would that work for your application?

The new feature of latched publishing would be useful, but I expect non-latched publishing in my application.
And I hesitate to change the current behavior of the node, so the new feature should be with conditional rosparam (ex. `~latched: True/False`), I think.

**Comment by paulbovbel** [comment]
I see the old while loop now, thanks. LGTM

**Comment by wkentaro** [comment]
Thanks, I squashed the commits.

**Comment by paulbovbel** [comment]
Going to need a corresponding PR in pcl_msgs

**Comment by wkentaro** [comment]
I have send PR ros-perception/pcl_msgs#5

**Comment by wkentaro** [comment]
pign

**Comment by wkentaro** [comment]
ping

**Comment by wkentaro** [comment]
ping

**Comment by paulbovbel** [comment]
Closing in favor of #213

**Pull request closed without merging**